text/0039-local-keys.md

```diff
@@ -0,0 +1,153 @@
+
+With this design decision, we will retain the possibility of adding a data migration in the future to read all (key, value) pairs through the index and move the data into an expanded space that will consist of a "URef" address and a hash of "key" bytes concatenated. We will programmatically replace the local keys' values with a pair that contains unchanged key bytes and a value itself during migration. The migration will remove a secondary index, and then we will rely on a prefix-based iteration mechanism in list_local.
+
+### Key::Local
+
+Adds a new variant that holds 32 bytes hashed value based on:
+
+```
+blake2b256(uref_addr || key_bytes)
+```
+
+Further operations such as read/write use the algorithm as mentioned earlier to access the `Key::Local` space.
+
+### create_local
+
+```rust
+fn create_local() -> URef;
+```
+
+Host provisions an URef of "unit" type under the hood used internally to store a secondary index of keys. This function's caller should keep the URef mentioned above under account or contract's named keys, retaining his ability to add new data.
+
+Newly created will have a type set to "unit" to hide the implementation details and discourage using "read" or "write" APIs to operate on local keyspace. The host will also secure this possibility on the backend, and inside of these host functions, a particular check will verify and forbid direct operations on the `Key::Local` space.
+
+### read_local
+
+```rust
+fn read_local(uref, key) -> T;
+```
+
+Reads data from a local key under "URef" and "key" bytes and returns a stored value. The system then verifies that URef is valid in the caller context and performs hashing of the "URef" address and the key bytes.
+
+A secondary key index is not involved in this operation.
+
+### write_local
+
+```rust
+fn write_local(uref, key, data);
+```
+
+Writes data under a key to a local key referenced by "URef.". The system then verifies that URef is in the execution context, and an additional check is in place to confirm if the "WRITE" permission bit is on.
+
+If the given "key" is not present in the secondary index that "URef" points to, then it is extended with "key." bytes.
+
+### add_local
+
+```rust
+fn add_local(uref, key, data);
+```
+
+Similar to `write_local,` but emits an `ADD` transformation for a given local key.
+
+### list_local
+
+```rust
+fn list_local(uref) -> Vec<Vec<u8>>;
+```
+
+This function lists all keys under a Key::Local referenced by "URef." Internally it reads all the keys from the secondary index and returns them to the caller.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+As explained above current limitation of key lengths inside a trie requires us to maintain a secondary index of keys under a URef to retain the ability to migrate into the unhashed, more extended keyspace for more efficient iteration.
```

Comment:
This restriction could be lifted.  @henrytill is it worth getting support for variable length keys so we can avoid maintaining this secondary index, or should that be done at a different time?